### PR TITLE
Improve management of credential keys

### LINKED ISSUE #134: Request: Add a function to delete all keys
There is also an bug where all my keys id stay the same.

### PR DESCRIPTION
**Pull request by rich-iannone**
This adds the `delete_credential_key()` and `delete_all_credential_keys()` functions, improves documentation on all existing functions dealing with credential keys, disallows creation of duplicate keys or keys with non-conforming names or types, and makes the associated error messages clearer to understand.

Fixes #134 
Fixes https://github.com/rich-iannone/blastula/issues/131
Fixes https://github.com/rich-iannone/blastula/issues/126
Fixes https://github.com/rich-iannone/blastula/issues/88
Fixes https://github.com/rich-iannone/blastula/issues/99

### CODE REVIEW

File: R/create_credentials.R [hunk 1]
```diff
@@ -120,6 +126,39 @@ create_smtp_creds_key <- function(id,
   # Determine whether the keyring package can be used
   validate_keyring_capable()
 
+  # Stop function if `id` provided is not of the right type
+  if (!(inherits(id, "character") | inherits(id, "numeric") | inherits(id, "integer"))) {
```
Comment (jcheng5) [inline R/create_credentials.R:130]: Use `||` instead of `|`

File: R/create_credentials.R [hunk 2]
```diff
@@ -120,6 +126,39 @@ create_smtp_creds_key <- function(id,
...
+  }
+
+  # Stop function if the `id` value isn't of length 1
+  if (length(id) != 1) {
+    stop("Only a vector of length 1 should be used for the `id` value.",
+         call. = FALSE)
+  }
+
+  # Stop function if the `id` is an empty string or NA
+  if (id == "" || is.na(id)) {
```
Comment (jcheng5) [inline R/create_credentials.R:142]: `id == "" || is.na(id)` should be `is.na(id) || id == ""`

File: R/create_credentials.R [hunk 3]
```diff
@@ -120,6 +126,39 @@ create_smtp_creds_key <- function(id,
...
+    stop("The value for `id` should not be `NA` or an empty string.",
+         call. = FALSE)
+  }
+
+  if (is.character(id) & grepl("-", id)) {
```
Comment (jcheng5) [inline R/create_credentials.R:147]: `&&`

File: R/create_credentials.R [hunk 4]
```diff
@@ -120,6 +126,39 @@ create_smtp_creds_key <- function(id,
...
+  }
+
+  # Determine whether the key already exists, stop if it does
+  creds_tbl <- get_keyring_creds_table()
+
+  if (id %in% creds_tbl$id) {
+    stop("The specified `id` corresponds to a credential key already in the key-value store:\n",
+         "* Use a different `id` value with `create_smtp_creds_key()`, or\n",
+         "* Delete the existing key with `delete_credential_key(id = \"", id,"\") and try again",
+         call. = FALSE)
```
Comment (jcheng5) [inline R/create_credentials.R:159]: It'd be nice to have an argument to this function that allows overwrite if it already exists.

File: R/retrieve_credentials.R [hunk 5]
```diff
@@ -1,13 +1,153 @@
...
 view_credential_keys <- function() {
 
+  # Viewing credentials that are on the system-wide key-value
+  # store requires the installation of the keyring package
+  if (!requireNamespace("keyring", quietly = TRUE)) {
+
+    stop("The `keyring` package is required for using the ",
+         "`view_credential_keys()` function",
+         call. = FALSE)
+  }
```
Comment (jcheng5) [inline R/retrieve_credentials.R:36]: DRY

File: R/retrieve_credentials.R [hunk 6]
```diff
@@ -1,13 +1,153 @@
...
+  if (!(id %in% ids_available)) {
+    stop("The specified `id` doesn't correspond to a credential key:\n",
+         "* Use the `view_credential_keys()` function to examine which `id` values are valid",
+         call. = FALSE)
+  }
+
+  # Obtain the matching service name that corresponds to the `id` value
+  key_to_delete <-
+    creds_tbl %>%
+    dplyr::filter(id == {{ id }}) %>%
```
Comment (jcheng5) [inline R/retrieve_credentials.R:92]: `!!id` instead of `{{ id }}`

File: R/retrieve_credentials.R [hunk 7]
```diff
@@ -1,13 +1,153 @@
...
+    dplyr::pull(key_name)
+
+  # Delete with `keyring::key_delete()`
+  keyring::key_delete(service = key_to_delete)
+
+  message(paste0("The blastula credential key with the `id` of \"", id, "\" has been deleted."))
```
Comment (jcheng5) [inline R/retrieve_credentials.R:98]: Let's not have success messages that don't indicate any useful information other than success.

File: R/retrieve_credentials.R (hunk 5 above)
Comment (rich-iannone) [reply R/retrieve_credentials.R:36]: Now using a util function called `validate_keyring_available()`.

File: R/create_credentials.R (hunk 4 above)
Comment (rich-iannone) [reply R/create_credentials.R:159]: Included the `overwrite` (default is `FALSE`). An associated message has been changed to inform users of this option.

File: R/retrieve_credentials.R (hunk 6 above)
Comment (rich-iannone) [reply R/retrieve_credentials.R]: Changed to use `!!` now.

File: R/retrieve_credentials.R (hunk 7 above)
Comment (rich-iannone) [reply R/retrieve_credentials.R:98]: In the two deletion functions, the message has been removed (with `invisible()` used instead).

File: R/create_credentials.R (hunk 3 above)
Comment (rich-iannone) [reply R/create_credentials.R:147]: Changed.

File: R/create_credentials.R (hunk 2 above)
Comment (rich-iannone) [reply R/create_credentials.R:142]: Made this change. Thanks.

File: R/create_credentials.R (hunk 1 above)
Comment (rich-iannone) [reply R/create_credentials.R:130]: Now changed. Thanks for the explanations earlier.

File: R/create_credentials.R [hunk 8]
```diff
@@ -228,7 +279,22 @@ creds_internal <- function(user = NULL,
   )
 }
 
-#' Stops function is the system is not capable of using \pkg{keyring}
+#' Stops function if the **keyring** package isn't installed
+#'
+#' @noRd
+validate_keyring_available <- function(fn_name) {
```
Comment (jcheng5) [inline R/create_credentials.R:285]: This function should also incorporate `validate_keyring_capable`

Comment (rich-iannone) [reply R/create_credentials.R]: Now done.

**Pull request merged**